File: Java_client/bim-portal-enhanced/src/main/java/com/pb40/bimportal/examples/ContextInfoExportExample.java
package com.pb40.bimportal.examples;

import com.bimportal.client.model.*;
import com.pb40.bimportal.client.BimPortalClientBuilder;
import com.pb40.bimportal.client.EnhancedBimPortalClient;
import com.pb40.bimportal.config.BimPortalConfig;
import com.pb40.bimportal.util.ExportUtils;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Context information export examples for BIM Portal Java client. Demonstrates context information
 * export workflows in multiple formats including PDF and OpenOffice with automatic file type
 * detection.
 */
public class ContextInfoExportExample {

  private static final Logger logger = LoggerFactory.getLogger(ContextInfoExportExample.class);

  /**
   * Check if credentials are available for authentication.
   *
   * @return True if credentials are configured
   */
  private static boolean checkCredentials() {
    if (!BimPortalConfig.hasCredentials()) {
      System.out.println("============================================================");
      System.out.println("WARNING: Credentials not found in environment variables.");
      System.out.println(
          "Please set "
              + BimPortalConfig.USERNAME_ENV_VAR
              + " and "
              + BimPortalConfig.PASSWORD_ENV_VAR
              + " in .env file.");
      System.out.println("============================================================");
      return false;
    }
    return true;
  }

  /**
   * Run context information export examples with the fixed search method.
   *
   * @param client Enhanced BIM Portal client
   */
  public static void runContextInfoExportExamples(EnhancedBimPortalClient client) {
    System.out.println("\n============================================================");
    System.out.println("📚 CONTEXT INFORMATION EXPORT EXAMPLES");
    System.out.println("============================================================");

    System.out.println("\n1️⃣ Searching for available context information...");
    try {
      List<SimpleContextInfoPublicDto> contextInfos = client.searchContextInfo();
      if (contextInfos.isEmpty()) {
        System.out.println("🔭 No context information found for export");
        return;
      }

      System.out.println("✅ Found " + contextInfos.size() + " context information items:");
      for (int i = 0; i < Math.min(5, contextInfos.size()); i++) {
        SimpleContextInfoPublicDto context = contextInfos.get(i);
        System.out.println(
            "   " + (i + 1) + ". " + context.getName() + " (" + context.getGuid() + ")");
      }

      SimpleContextInfoPublicDto selectedContext = contextInfos.getFirst();
      System.out.println("\n🎯 Using context info: '" + selectedContext.getName() + "'");

      Map<String, Path> exportResults = new HashMap<>();

      System.out.println("\n2️⃣ Exporting context information...");

      // PDF Export with auto-detection
      System.out.println("   📄 Exporting as PDF...");
      Optional<byte[]> pdfContent = client.exportContextInfoPdf(selectedContext.getGuid());
      if (pdfContent.isPresent()) {
        String baseFilename = "context_pdf_" + selectedContext.getGuid();
        Optional<Path> pdfPath =
            ExportUtils.exportWithDetection(pdfContent.get(), baseFilename, "pdf");
        if (pdfPath.isPresent()) {
          exportResults.put("PDF", pdfPath.get());
          System.out.println("   ✅ PDF exported: " + pdfPath.get());
        } else {
          System.out.println("   ❌ PDF export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ PDF export failed: No content received");
      }

      // OpenOffice Export with auto-detection
      System.out.println("   📝 Exporting as OpenOffice...");
      Optional<byte[]> odtContent = client.exportContextInfoOpenOffice(selectedContext.getGuid());
      if (odtContent.isPresent()) {
        String baseFilename = "context_odt_" + selectedContext.getGuid();
        Optional<Path> odtPath =
            ExportUtils.exportWithDetection(odtContent.get(), baseFilename, "odt");
        if (odtPath.isPresent()) {
          exportResults.put("OpenOffice", odtPath.get());
          System.out.println("   ✅ OpenOffice exported: " + odtPath.get());
        } else {
          System.out.println("   ❌ OpenOffice export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ OpenOffice export failed: No content received");
      }

      // Summary with file type information
      System.out.println("\n📈 Export Summary: " + exportResults.size() + "/2 formats successful");
      for (Map.Entry<String, Path> entry : exportResults.entrySet()) {
        String filename = entry.getValue().getFileName().toString();
        String extension = filename.substring(filename.lastIndexOf('.') + 1).toUpperCase();
        System.out.println("   ✅ " + entry.getKey() + " (" + extension + "): " + entry.getValue());
      }

      // Additional context info details
      System.out.println("\n📋 Context Information Details:");
      try {
        Optional<AIAContextInfoPublicDto> contextDetails =
            client.getContextInfo(selectedContext.getGuid());
        if (contextDetails.isPresent()) {
          AIAContextInfoPublicDto details = contextDetails.get();
          System.out.println("   Name: " + details.getName());
          System.out.println("   GUID: " + details.getGuid());
          System.out.println(
              "   Description: "
                  + (details.getDescription() != null ? details.getDescription() : "N/A"));
        } else {
          System.out.println("   Could not retrieve detailed information");
        }
      } catch (Exception e) {
        System.out.println("   Error retrieving details: " + e.getMessage());
      }

    } catch (Exception e) {
      logger.error("❌ Error in context info export examples", e);
      System.err.println("❌ Error in context info export examples: " + e.getMessage());
    }
  }

  /** Main method to run context information export examples. */
  public static void main(String[] args) {
    System.out.println("======================================================================");
    System.out.println("🚀 BIM PORTAL CONTEXT INFORMATION EXPORT EXAMPLES");
    System.out.println("======================================================================");

    if (!checkCredentials()) {
      System.err.println("❌ Cannot run export examples without credentials");
      return;
    }

    if (!ExportUtils.isExportDirectoryWritable()) {
      System.err.println("❌ Export directory is not writable!");
      return;
    }

    System.out.println("🔧 Setting up authenticated client...");

    try {
      EnhancedBimPortalClient client = BimPortalClientBuilder.buildDefault();

      EnhancedBimPortalClient.HealthCheckResult healthCheck = client.performHealthCheck();
      System.out.println("🩺 Health check: " + healthCheck);

      if (!healthCheck.isApiAccessible()) {
        System.err.println("❌ API is not accessible. Please check configuration.");
        return;
      }

      // Run the fixed context info export examples
      runContextInfoExportExamples(client);

      System.out.println(
          "\n======================================================================");
      System.out.println("✅ CONTEXT INFORMATION EXPORT EXAMPLES COMPLETE!");
      System.out.println("======================================================================");
      System.out.println(
          "📂 Check the '" + BimPortalConfig.getExportPath() + "' directory for exported files");

      client.logout();
      System.out.println("👋 Logged out and session closed.");

    } catch (Exception e) {
      logger.error("❌ Error running context information export examples", e);
      System.err.println("❌ Error running examples: " + e.getMessage());
    }
  }
}

File: Python_client/examples/export_examples/context_info_export_example.py
"""
Context information export examples for BIM Portal Python client.

This module demonstrates context information export workflows in multiple formats including 
PDF and OpenOffice with automatic file type detection.
"""

import os
import sys
from pathlib import Path

# Ensure we can import from project root
project_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(project_root))

import logging
from typing import Dict, Optional

from dotenv import load_dotenv

from client.auth.auth_config import BIM_PORTAL_PASSWORD_ENV_VAR, BIM_PORTAL_USERNAME_ENV_VAR
from client.auth.auth_service_impl import AuthService
from client.enhanced_bim_client import EnhancedBimPortalClient
from client.config import BIMPortalConfig
from examples.export_examples.utils.common_utils import check_credentials
from examples.export_examples.utils.export_utils import ExportUtils

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Load environment variables
load_dotenv()



def run_context_info_export_examples(client: EnhancedBimPortalClient):
    """
    Run context information export examples with comprehensive format support.
    
    Args:
        client: Enhanced BIM Portal client
    """
    print("\n" + "=" * 60)
    print("📚 CONTEXT INFORMATION EXPORT EXAMPLES")
    print("=" * 60)
    
    print("\n1️⃣ Searching for available context information...")
    try:
        context_infos = client.search_context_info()
        if not context_infos:
            print("🔭 No context information found for export")
            return
        
        print(f"✅ Found {len(context_infos)} context information items:")
        for i, context in enumerate(context_infos[:5], 1):
            print(f"   {i}. {context.name} ({context.guid})")
        
        selected_context = context_infos[0]
        print(f"\n🎯 Using context info: '{selected_context.name}'")
        
        export_results: Dict[str, Optional[Path]] = {}
        
        print("\n2️⃣ Exporting context information...")
        
        # PDF Export with auto-detection
        print("   📄 Exporting as PDF...")
        pdf_content = client.export_context_info_pdf(selected_context.guid)
        if pdf_content:
            base_filename = f"context_pdf_{selected_context.guid}"
            pdf_path = ExportUtils.export_with_detection(pdf_content, base_filename, "pdf")
            if pdf_path:
                export_results['PDF'] = pdf_path
                print(f"   ✅ PDF exported: {pdf_path}")
            else:
                print("   ❌ PDF export failed: Could not save file")
                export_results['PDF'] = None
        else:
            print("   ❌ PDF export failed: No content received")
            export_results['PDF'] = None
        
        # OpenOffice Export with auto-detection
        print("   📝 Exporting as OpenOffice...")
        odt_content = client.export_context_info_openoffice(selected_context.guid)
        if odt_content:
            base_filename = f"context_odt_{selected_context.guid}"
            odt_path = ExportUtils.export_with_detection(odt_content, base_filename, "odt")
            if odt_path:
                export_results['OpenOffice'] = odt_path
                print(f"   ✅ OpenOffice exported: {odt_path}")
            else:
                print("   ❌ OpenOffice export failed: Could not save file")
                export_results['OpenOffice'] = None
        else:
            print("   ❌ OpenOffice export failed: No content received")
            export_results['OpenOffice'] = None
        
        # Summary with file type information
        successful_exports = len([path for path in export_results.values() if path is not None])
        total_formats = len(export_results)
        
        print(f"\n📈 Export Summary: {successful_exports}/{total_formats} formats successful")
        
        for format_name, path in export_results.items():
            if path:
                extension = path.suffix[1:].upper() if path.suffix else "UNKNOWN"
                print(f"   ✅ {format_name} ({extension}): {path}")
            else:
                print(f"   ❌ {format_name}: Export failed")
        
        # Additional context info details
        print(f"\n📋 Context Information Details:")
        try:
            detailed_context = client.get_context_info(selected_context.guid)
            if detailed_context:
                print(f"   Name: {detailed_context.name}")
                print(f"   GUID: {detailed_context.guid}")
                if hasattr(detailed_context, 'description') and detailed_context.description:
                    print(f"   Description: {detailed_context.description}")
                if hasattr(detailed_context, 'version') and detailed_context.version:
                    print(f"   Version: {detailed_context.version}")
                if hasattr(detailed_context, 'context_type') and detailed_context.context_type:
                    print(f"   Type: {detailed_context.context_type}")
            else:
                print("   Could not retrieve detailed information")
        except Exception as e:
            print(f"   Error retrieving details: {e}")
        
        if successful_exports == 0:
            print("💡 Note: Context information may require special permissions to export")
        
    except Exception as e:
        logger.error(f"Error in context info export examples: {e}")
        print(f"❌ Error in context info export examples: {e}")


def main():
    """Main method to run context information export examples."""
    print("=" * 70)
    print("🚀 BIM PORTAL CONTEXT INFORMATION EXPORT EXAMPLES")
    print("=" * 70)
    
    if not check_credentials():
        print("❌ Cannot run export examples without credentials")
        return
    
    if not ExportUtils.is_export_directory_writable():
        print("❌ Export directory is not writable!")
        return
    
    print("🔧 Setting up authenticated client...")
    
    try:
        auth_service = AuthService()
        client = EnhancedBimPortalClient(
            auth_service=auth_service, 
            base_url=BIMPortalConfig.BASE_URL
        )
        
        # Run context info export examples
        run_context_info_export_examples(client)
        
        print("\n" + "=" * 70)
        print("✅ CONTEXT INFORMATION EXPORT EXAMPLES COMPLETE!")
        print("=" * 70)
        print(f"📂 Check the '{BIMPortalConfig.EXPORT_DIRECTORY}' directory for exported files")
        
    except Exception as e:
        logger.error(f"Error running context information export examples: {e}")
        print(f"❌ Error running examples: {e}")


if __name__ == "__main__":
    main()
